File: examples/telemetry/10_performance_metrics.py
#!/usr/bin/env python3
# SPDX-FileCopyrightText: Copyright (C) 2026 MindTenet LLC
# SPDX-License-Identifier: Apache-2.0
# SPDX-Comment: Part of Undef Telemetry.
#

"""⚡ Performance characteristics of the telemetry library.

Demonstrates:
- Import time of the full undef.telemetry package
- configure_logging() cost
- Hot-path instrument ops: counter.add(), gauge.set(), histogram.record()
- Sampling decision throughput via should_sample()
- Event name construction via event_name()
- Full setup_telemetry() / shutdown_telemetry() lifecycle cost
- Lazy-loading verification: processors.py has no direct slo dependency
"""

from __future__ import annotations

import ast
import time
from pathlib import Path


def _bench(fn: object, iterations: int = 10_000) -> float:
    """Run *fn* for *iterations* and return ns/op."""
    callable_fn = fn  # type: ignore[assignment]
    start = time.perf_counter_ns()
    for _ in range(iterations):
        callable_fn()
    elapsed_ns = time.perf_counter_ns() - start
    ns_per_op = elapsed_ns / iterations
    return ns_per_op


def _fmt(ns: float) -> str:
    if ns >= 1_000_000:
        return f"{ns / 1_000_000:>10.2f} ms"
    if ns >= 1_000:
        return f"{ns / 1_000:>10.2f} us"
    return f"{ns:>10.0f} ns"


def main() -> None:
    print("⚡ Performance Characteristics\n")

    rows: list[tuple[str, str]] = []

    # ── 📦 Full package import ───────────────────────────────────────
    print("📦 Full Package Import\n")
    t0 = time.perf_counter_ns()
    __import__("undef.telemetry")
    t1 = time.perf_counter_ns()
    rows.append(("import undef.telemetry", _fmt(t1 - t0)))

    # ── ⚙️  configure_logging ────────────────────────────────────────
    print("⚙️  Logging Configuration\n")
    from undef.telemetry.config import TelemetryConfig
    from undef.telemetry.logger.core import configure_logging

    cfg = TelemetryConfig()

    def do_configure() -> None:
        configure_logging(cfg)

    rows.append(("configure_logging()", _fmt(_bench(do_configure, iterations=100))))

    # ── 🔥 Hot-path ops ──────────────────────────────────────────────
    print("🔥 Hot-Path Instrument Operations\n")
    from undef.telemetry import counter, event_name, gauge, histogram
    from undef.telemetry.sampling import should_sample

    c = counter("perf.example.requests", "bench counter")
    g = gauge("perf.example.active", "bench gauge")
    h = histogram("perf.example.latency", "bench histogram", unit="ms")

    rows.append(("counter.add(1)", _fmt(_bench(lambda: c.add(1)))))
    rows.append(("gauge.set(42)", _fmt(_bench(lambda: g.set(42)))))
    rows.append(("histogram.record(3.14)", _fmt(_bench(lambda: h.record(3.14)))))
    rows.append(
        (
            'should_sample("logs", "x")',
            _fmt(_bench(lambda: should_sample("logs", "perf.test"))),
        )
    )
    rows.append(
        (
            'event_name("a","b","c")',
            _fmt(_bench(lambda: event_name("perf", "bench", "op"))),
        )
    )

    # ── 🔄 Full lifecycle ────────────────────────────────────────────
    print("🔄 Setup / Shutdown Lifecycle\n")
    from undef.telemetry import setup_telemetry, shutdown_telemetry

    shutdown_telemetry()

    def lifecycle() -> None:
        setup_telemetry()
        shutdown_telemetry()

    rows.append(("setup + shutdown cycle", _fmt(_bench(lifecycle, iterations=50))))

    # ── 📊 Results table ─────────────────────────────────────────────
    print("📊 Results\n")
    max_label = max(len(r[0]) for r in rows)
    for label, value in rows:
        print(f"    {label:<{max_label}}  {value}")

    # ── 🔌 Lazy-loading verification ─────────────────────────────────
    print("\n🔌 Lazy-Loading Verification\n")
    print("    processors.py uses a lazy import for classify_error (from slo.py)")
    print("    so it has no direct module-level dependency on the metrics subsystem.")
    print("    Verifying via AST analysis...\n")

    from undef.telemetry.logger import processors as proc_mod

    src = Path(str(proc_mod.__file__)).read_text()
    tree = ast.parse(src)
    slo_at_top = any(isinstance(stmt, ast.ImportFrom) and stmt.module and "slo" in stmt.module for stmt in tree.body)
    if slo_at_top:
        print("    ❌ slo is imported at module level (eager)")
    else:
        print("    ✅ No top-level slo import in processors.py (lazy, deferred to call time)")

    print("\n🏁 Done!")


if __name__ == "__main__":
    main()

File: typescript/examples/telemetry/10_performance_metrics.ts
// SPDX-FileCopyrightText: Copyright (C) 2026 provide.io llc
// SPDX-License-Identifier: Apache-2.0
// SPDX-Comment: Part of Provide Telemetry.

/**
 * ⚡ Performance characteristics of the telemetry library.
 *
 * Demonstrates:
 * - Import time of the full @provide-io/telemetry package (measured at startup)
 * - setupTelemetry() cost
 * - Hot-path instrument ops: counter.add(), gauge.add(), histogram.record()
 * - Sampling decision throughput via shouldSample()
 * - Event name construction via eventName()
 * - Full setup / shutdown lifecycle cost
 * - Import footprint: no eager OTEL SDK pull
 *
 * Run: npx tsx examples/telemetry/10_performance_metrics.ts
 */

import {
  counter,
  eventName,
  gauge,
  histogram,
  setupTelemetry,
  shouldSample,
  shutdownTelemetry,
} from '../../src/index.js';

const ITERATIONS = 10_000;
const LIFECYCLE_ITERATIONS = 50;

function bench(fn: () => void, iterations: number = ITERATIONS): number {
  const start = performance.now();
  for (let i = 0; i < iterations; i++) fn();
  return ((performance.now() - start) * 1_000_000) / iterations; // ns per op
}

function fmt(ns: number): string {
  if (ns >= 1_000_000) return `${(ns / 1_000_000).toFixed(2).padStart(10)} ms`;
  if (ns >= 1_000) return `${(ns / 1_000).toFixed(2).padStart(10)} µs`;
  return `${ns.toFixed(0).padStart(10)} ns`;
}

async function main(): Promise<void> {
  console.log('⚡ Performance Characteristics\n');

  const rows: [string, string][] = [];

  // ── ⚙️  setupTelemetry cost ──────────────────────────────
  console.log('⚙️  Setup Cost\n');
  rows.push(['setupTelemetry()', fmt(bench(() => setupTelemetry({ serviceName: 'perf-test' }), 100))]);

  setupTelemetry({ serviceName: 'perf-test', consoleOutput: false });

  // ── 🔥 Hot-path ops ──────────────────────────────────────
  console.log('🔥 Hot-Path Instrument Operations\n');
  const c = counter('perf.example.requests', { description: 'bench counter' });
  const g = gauge('perf.example.active', { description: 'bench gauge' });
  const h = histogram('perf.example.latency', { description: 'bench histogram', unit: 'ms' });

  rows.push(['counter.add(1)', fmt(bench(() => c.add(1)))]);
  rows.push(['gauge.add(42)', fmt(bench(() => g.add(42)))]);
  rows.push(['histogram.record(3.14)', fmt(bench(() => h.record(3.14)))]);
  rows.push(['shouldSample("logs")', fmt(bench(() => shouldSample('logs')))]);
  rows.push(['eventName("a","b","c")', fmt(bench(() => eventName('perf', 'bench', 'op')))]);

  // ── 🔄 Full lifecycle ────────────────────────────────────
  console.log('🔄 Setup / Shutdown Lifecycle\n');
  rows.push([
    'setup + shutdown cycle',
    fmt(
      bench(async () => {
        setupTelemetry({ serviceName: 'perf-test' });
        await shutdownTelemetry();
      }, LIFECYCLE_ITERATIONS),
    ),
  ]);

  // ── 📊 Results table ─────────────────────────────────────
  console.log('📊 Results\n');
  const maxLabel = Math.max(...rows.map(([label]) => label.length));
  for (const [label, value] of rows) {
    console.log(`    ${label.padEnd(maxLabel)}  ${value}`);
  }

  // ── 🔌 Import footprint check ─────────────────────────────
  console.log('\n🔌 Import Footprint\n');
  // ESM: use process.moduleLoadList as a proxy for loaded modules
  const loadedMods = (process.moduleLoadList ?? []).filter((k: string) => k.includes('provide-telemetry') || k.includes('@provide-io'));
  console.log(`    Loaded @provide-io/telemetry source files: ${loadedMods.length}`);
  const hasOtelSdk = loadedMods.some((k: string) => k.includes('sdk-trace') || k.includes('sdk-metrics'));
  if (hasOtelSdk) {
    console.log('    ⚠️  OTEL SDK loaded (expected when peer deps are installed)');
  } else {
    console.log('    ✅ OTEL SDK not loaded at import time (lazy / noop API only)');
  }

  console.log('\n🏁 Done!');
  await shutdownTelemetry();
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
